moshi/src/main/java/com/squareup/moshi/Moshi.kt

```diff
@@ -0,0 +1,388 @@
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.moshi.Types.createJsonQualifierImplementation
+import com.squareup.moshi.internal.Util
+import com.squareup.moshi.internal.Util.canonicalize
+import com.squareup.moshi.internal.Util.isAnnotationPresent
+import com.squareup.moshi.internal.Util.removeSubtypeWildcard
+import com.squareup.moshi.internal.Util.typeAnnotatedWithAnnotations
+import com.squareup.moshi.internal.Util.typesMatch
+import java.lang.reflect.Type
+import javax.annotation.CheckReturnValue
+
+/**
+ * Coordinates binding between JSON values and Java objects.
+ *
+ * Moshi instances are thread-safe, meaning multiple threads can safely use a single instance
+ * concurrently.
+ */
+public class Moshi internal constructor(builder: Builder) {
+  private val factories: List<JsonAdapter.Factory> = buildList {
+    addAll(builder.factories)
+    addAll(BUILT_IN_FACTORIES)
+  }
+  private val lastOffset: Int = builder.lastOffset
+  private val lookupChainThreadLocal = ThreadLocal<LookupChain>()
+  private val adapterCache: MutableMap<Any?, JsonAdapter<*>?> = LinkedHashMap()
+
+  /** Returns a JSON adapter for `type`, creating it if necessary. */
+  @CheckReturnValue
+  public fun <T> adapter(type: Type): JsonAdapter<T> {
+    return adapter(type, Util.NO_ANNOTATIONS)
+  }
+
+  @CheckReturnValue
+  public fun <T> adapter(type: Class<T>): JsonAdapter<T> {
+    return adapter(type, Util.NO_ANNOTATIONS)
+  }
+
+  @CheckReturnValue
+  public fun <T> adapter(type: Type, annotationType: Class<out Annotation>): JsonAdapter<T> {
+    return adapter(type, setOf(createJsonQualifierImplementation(annotationType)))
+  }
+
+  @CheckReturnValue
+  public fun <T> adapter(type: Type, vararg annotationTypes: Class<out Annotation>): JsonAdapter<T> {
+    if (annotationTypes.size == 1) {
+      return adapter(type, annotationTypes[0])
+    }
+    val annotations = buildSet(annotationTypes.size) {
+      for (annotationType in annotationTypes) {
+        add(createJsonQualifierImplementation(annotationType)!!)
+      }
+    }
+    return adapter(type, annotations)
+  }
+
+  @CheckReturnValue
+  public fun <T> adapter(type: Type, annotations: Set<Annotation>): JsonAdapter<T> {
+    return adapter(type, annotations, fieldName = null)
+  }
+
+  /**
+   * @param fieldName An optional field name associated with this type. The field name is used as a
+   * hint for better adapter lookup error messages for nested structures.
+   */
+  @CheckReturnValue // Factories are required to return only matching JsonAdapters.
```

Comment:
It seems that here and on the next method this comment was attached to `@SuppressWarnings` and not `@CheckReturnValue`. If `@SuppressWarnings` is not needed anymore, should we remove the comment?

Reply:
Moved in 70091a9